src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,441 @@
+                fmt = identify_file_format(k)
+                if fmt == "root":
+                    if "format" not in v:
+                        v["format"] = "root"
+                    else:
+                        assert (
+                            v["format"] == "root"
+                        ), f"Expected 'format' to be 'root', got {v['format']} for {k}"
+                elif fmt == "parquet":
+                    if "format" not in v:
+                        v["format"] = "parquet"
+                    else:
+                        assert (
+                            v["format"] == "parquet"
+                        ), f"Expected 'format' to be 'parquet', got {v['format']} for {k}"
+        return data
+
+
+class InputFiles(
+    RootModel[
+        dict[
+            str,
+            Union[
+                CoffeaROOTFileSpec,
+                CoffeaParquetFileSpec,
+                CoffeaROOTFileSpecOptional,
+                CoffeaParquetFileSpecOptional,
+            ],
+        ]
+    ],
+    DictMethodsMixin,
+    InputFilesMixin,
+):
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.root)
+
+    @model_validator(mode="after")
+    def promote_and_check_files(self) -> Self:
+        preprocessed = {k: False for k in self.root}
+        for k, v in self.root.items():
+            try:
+                if type(v) in [CoffeaROOTFileSpecOptional]:
+                    self.root[k] = CoffeaROOTFileSpec(v)
+                if type(v) in [CoffeaParquetFileSpecOptional]:
+                    self.root[k] = CoffeaParquetFileSpec(v)
+                preprocessed[k] = True
+            except Exception:
+                pass
+        if all(preprocessed.values()):
+            try:
+                self.root = PreprocessedFiles(self.root).root
+            except Exception:
+                pass
+        return self
+
+
+class PreprocessedFiles(
+    RootModel[
+        dict[
+            str,
+            Union[
+                CoffeaROOTFileSpec,
+                CoffeaParquetFileSpec,
+            ],
+        ]
+    ],
+    DictMethodsMixin,
+    InputFilesMixin,
+):
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.root)
+
+    @model_validator(mode="after")
+    def promote_and_check_files(self) -> Self:
+        preprocessed = {k: False for k in self.root}
+        for k, v in self.root.items():
+            try:
+                if type(v) in [CoffeaROOTFileSpecOptional]:
+                    self.root[k] = CoffeaROOTFileSpec(v)
+                if type(v) in [CoffeaParquetFileSpecOptional]:
+                    self.root[k] = CoffeaParquetFileSpec(v)
+                preprocessed[k] = True
+            except Exception:
+                pass
+        if all(preprocessed.values()):
+            return self
+
+
+class DatasetSpec(BaseModel):
+    files: InputFiles
+    metadata: dict[Hashable, Any] | None = None
+    format: str | None = None
+    form: str | None = None
```

Comment:
Change to `form_blob` or `compressed_form` or something and make `form` a property